src/memory.rs

```diff
@@ -432,13 +392,13 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
     }
     
     // Returns a dtor and its argument, if one is supposed to run
-    pub(crate) fn fetch_tls_dtor(&mut self) -> Option<(ty::Instance<'tcx>, Pointer)> {
-        for (_, &mut TlsEntry { ref mut data, dtor }) in self.thread_local.iter_mut() {
-            if !data.is_null_ptr() {
-                if let Some(dtor) = dtor {
-                    let old_data = *data;
-                    *data = Pointer::from_int(0);
-                    return Some((dtor, old_data));
+    pub(crate) fn fetch_tls_dtor(&mut self) -> Option<(ty::Instance<'tcx>, PrimVal)> {
+        for (_, &mut TlsEntry { ref mut data, ref mut dtor }) in self.thread_local.iter_mut() {
+            if let Some(dtor) = dtor.take() {
```

Comment:
I would think it gets added to the first round, but could be convinced otherwise.^^

I imagine something like the following (horribly inefficient) pseudocode:
```
while there_is_any_key_with_non_null_dtor_and_data() {
  let mut i = 0;
  while i < keys.len() {
    keys[i].run_dtor_if_non_null_data();
    i += 1;
  }
}
```